app/config.go

```diff
@@ -173,54 +184,76 @@ func loadExternalProperties() (map[string]interface{}, error) {
 			// Override through P1=V1,P2=V2
 			for _, pair := range strings.Split(propFile, ",") {
 				kv := strings.Split(pair, "=")
-				if len(kv) == 2 && kv[0] != "" &&  kv[1] != "" {
+				if len(kv) == 2 && kv[0] != "" && kv[1] != "" {
 					props[kv[0]] = kv[1]
 				} else {
 					logger.Warnf("'%s' is not valid override value. It must be in PropName=PropValue format.", pair)
 				}
 			}
 		}
+	}
 
-		if len(props) > 0 {
-			// Resolve property values
-			resolverType := config.GetAppPropertiesValueResolver()
-			if resolverType != "" {
-				logger.Infof("'%s' is set to '%s'. ", config.ENV_APP_PROPERTY_RESOLVER_KEY, resolverType)
-				resolver := GetPropertyValueResolver(resolverType)
-				if resolver == nil {
-					errMag := fmt.Sprintf("Unsupported resolver type - %s. Resolver not registered.", resolverType)
-					return nil, errors.New(errMag)
-				}
+	resolverType := config.GetAppPropertiesValueResolver()
+	var resolver PropertyValueResolver
+	if resolverType != "" {
+		logger.Infof("'%s' is set to '%s'. ", config.ENV_APP_PROPERTY_RESOLVER_KEY, resolverType)
+		resolver = GetPropertyValueResolver(resolverType)
+		if resolver == nil {
+			errMag := fmt.Sprintf("Unsupported resolver type - %s. Resolver not registered.", resolverType)
+			return nil, errors.New(errMag)
+		}
 
-				for k, v := range  props {
-					strVal, ok := v.(string)
-					if ok && len(strVal) > 0 {
-						if strVal[0] == '$' {
-							// Use resolver
-							newVal, err := resolver.ResolveValue(strVal[1:])
-							if err != nil {
-								return nil, err
-							}
-							props[k] = newVal
-
-							// May be a secret??
-							strVal, _ = newVal.(string)
+		if len(props) > 0 {
```

Comment:
The override app property might only override some of the properties, not all properties so we should use the resolver to resolve the rest of the app properties.

Reply:
@lixingwang : I don't think we should mix them. You can use either approach to override values. At this point, I dont see use cases where one would like override some using override mechanism and some using external config.

Reply:
@vijaynalawade Take @debovema 's case as an example.  he defined 2 properties `flogo/simple-config/message1` and `flogo/simple-config/message2` in flogo.json.

Now he only wants to override `flogo/simple-config/message1` so he specifies the property in AppPropertiesOverride, but the property `flogo/simple-config/message2` will not be resolved, right?  but it should be resolved.

Reply:
Just to understand well, what is the point of resolving environment variable in overridden properties' values ?
In current behaviour, if I define a property "property1" in flogo.json, I must set an environment variable FLOGO_APP_PROPS_OVERRIDE="property1=$PROPERTY1" and then set another environment variable PROPERTY1 with the actual value of "property1". That would be simpler to directly set an environment variable "property1" directly wouldn't it ?

Reply:
A proposal : why not making a resolver for each use case without giving the possibility to mix them ?

* the "file" resolver would use an external JSON properties file. The implementation of this resolver would be externalized in another Go file as I did for the Consul resolver (https://github.com/square-it/flogo-config-mgmt/blob/simple-consul-kv/consul/consul.go). The location would be set by an environment variable in this file and the file itself would be loaded in the ```init``` method and the ```ResolveValue``` method gets called for each property.
* the "env" resolver would try to resolve any property defined in flogo.json with the value of the environment variable of the same name (or maybe capitalized version of the property name). Implementation would be externalized as for "file". We could allow some naming pattern before calling ```os.Getenv```. For instance ```my_property_1``` would match ```my_property_1``` but also ```MY_PROPERTY_1```.
* other external resolver like "consul" one can be implemented too.

In the future we could also think of giving a list of resolver with priorities. For instance: ```FLOGO_APP_PROPS_VALUE_RESOLVER="consul,env,file"```

Reply:
@debovema :The basic idea behind FLOGO_APP_PROPS_OVERRIDE was to let developers override prop values without worrying about external configurations. I personally dont think we should over complicate this by forcing developers configure yet another variable to set the resolver type.

You can always configure $env[ENV_NAME] as a value to your app prop that should be resolved from an env var. Though, for the consistency with other resolvers like consul, we can certainly enhance env resolver to resolve props based on app prop name e.g. MyProp1 prop would be resolved to MyProp1 env var and C1.C2.MyProp2 to C1_C2_MyProp2. @mellistibco @fm-tibco ??

Earlier, we had a brief chat about chaining of resolvers but could not find any compelling use case to support it.

Reply:
My idea is not to add another property to set the resolver but rather to normalize the resolution of properties with the help of resolvers.

# Currently

Let's sum up what we have now:

* Override properties with a JSON-formatted file: definitely a must-have
* Override with a comma-separated list of key-value. It is interesting when only a property (or only a few) needs to be overridden.
However we found that it is quite tedious to set all these key-value pairs in an environment variable. It is not optimal for readability.
And at last the mechanism to redefine the value of one key-value pair to an environment variable which will be eventually resolved (by what you called a Resolver) is hard to understand (see issue https://github.com/TIBCOSoftware/flogo/issues/440 which is over-complicating the thing IMHO).

# Target (open to discussion)

What would be great:

* Add possibility to resolve properties from an external tool (Consul, etcd, zookeeper) in a microservices fashion (the original goal of this PR)
* Keep the possibility to provide a JSON file (very close to *profile*s in BW6): it is perfect to define an environment per file (Dev, Integration, Prod, ...) as clients are used to do
* Refine the override mechanism of properties through environment variables ([see below](#automatic-properties-exposition))
* Mix the different patterns in a **simple yet powerful way** with priorities

# To deprecate

What could/should be deprecated:

* The use of $env[MY_VAR] pattern should be deprecated. I understand it is convenient to begin a simple development.
However it will not scale at all and will be error-prone since these variables are not declared as properties are.
Moreover the mechanism of [automatic properties exposition](#automatic-properties-exposition) makes this use case obsolete.

# Automatic properties exposition

Let's dig into the details concerning the mechanism of properties override through environment variables.

I think that all declared properties in the flogo.json shall be automatically *exposed* as a corresonding environment variable.

By doing this, we will **leverage a cloud-native handling of Flogo application properties with the underlying container management software**: simple ```-e``` flag of ```docker run``` command, [environment variables of a Docker Compose file](https://docs.docker.com/compose/environment-variables/), [Kubernetes environment variables](https://kubernetes.io/docs/tasks/inject-data-application/define-environment-variable-container/), ...
The downside is that the properties can have different case convention (CamelCase, UPPERCASE_WITH_UNDERSCORE, ...). This can be mitigated by resolving the actual name (```MyProperty```) and a canonical one (MY_PROPERTY).
As I said above, the Flogo CLI could be updated to display the canonical form for each property or, even better, this could be included in the application itself with a ```--help``` switch !

Lastly, this allows to deprecate the $env[MY_VAR] syntax.

# Conclusion

Another constraint, as you said, is not to over-complicate the configuration. Let's define "env" and "file" resolvers as builtin ones.
And multiple resolvers would be implicitly defined and hard-coded as follows (by decreasing order of priority):
```
env (always resolved unless explicitly disabled) > file (none if not set) > external (Consul for instance, none if not set) > flogo.json (default value)
```

So the idea of chaining is used to promote convention-over-configuration paradigm:
* "env" is always used (top priority), if needed we could think of a flag to disable it ```FLOGO_APP_RESOLVE_ENV_VARS=false```. All properties are automatically *exposed* as environment variables.
* "file" is used if the right environment variable is set, ignored otherwise. I suggest ```FLOGO_APP_PROFILES="file1.json,file2.json"```. Notice the comma-separated list to manage inheritance of different profiles (something between a nice-to-have and a must-have).
* "consul" (or other *external* resolvers) can be set with the right environment variable.

The number of environment variables to configure stays low:
* 0, instead of 1 currently, to override some properties on the command line (since we directly define the overridden property as an environment variable). ```FLOGO_APP_PROPS_OVERRIDE``` is not used anymore for this feature. ```FLOGO_APP_RESOLVE_ENV_VARS``` is optional to disable this resolution.
* 1, instead of 1 currently, to override properties from file(s), ```FLOGO_APP_PROPS_OVERRIDE``` should be replaced by ```FLOGO_APP_PROFILES```.
* 1, instead of 1 currently, for external tool, ```FLOGO_APP_PROPS_VALUE_RESOLVER``` should be replaced by ```FLOGO_APP_EXTERNAL_CONFIG```.

I can implement this specification on Wednesday. With examples :)

Regards,

Mathieu

Reply:
@debovema Thanks for the detailed summary.
Can we split this PR into two or more? 
Let's keep this PR for original proposal of using app prop names in the resolution. 
Let's discuss other improvements in other PRs.

Reply:
I understand that we should limit the impacts by splitting modifications in different PRs.

What you point out about separation of concerns ("Sharing of configuration across activities/triggers" vs "Externalization of configuration") could be implemented with properties with visibility (public/private) but is not implemented (yet) in this PR.

I updated this PR with the minimum requirements from above "spec".

In fact it now use app prop names in the resolution and existing resolver has been updated to take this into account.

And I also updated the demo project : https://github.com/square-it/flogo-config-mgmt/blob/simple-consul-kv/consul/README.md

This demo has more testing with different use cases and shows the order priorities.

It describes the whole mechanism as it should be in the official Flogo documentation at last (extending https://tibcosoftware.github.io/flogo/development/flows/property-bag/ or creating a new page).